pkg/services/authz/rbac/service.go

```diff
@@ -640,7 +609,13 @@ func buildFolderList(scopes map[string]bool, tree map[string]FolderNode) *authzv
 		}
 
 		itemSet[identifier] = struct{}{}
-		getChildren(tree, identifier, itemSet)
+		for n := range tree.Children(identifier) {
+			if _, ok := itemSet[n.UID]; ok {
+				// we have already walked this part of the tree
+				break
+			}
```

Comment:
I'm trying to figure out if it's ok to break here. What if we have a tree something like this

```
    F
 S1   S2
NS1   NS2
```

Let's say the user has direct permissions on S1 and F, and we process S1 first - we will now have S1 and NS1 in the item set. Then we process F. We'll start with a queue of S1 and S2, and because S1 is already processed, we will break and never process S2 and NS2.

I think we also had the bug previously - we should've used `continue` instead of `return`.

But I also haven't tested this properly, will give it a go after the review.

Reply:
Yup, this scenario breaks. I've extended a test to show it: https://github.com/grafana/grafana/pull/99554/commits/38be4144822140b671b2f26ad402c4da29540371

I think we just need to remove the break statements for both places where we iterate over children. I can't think of a neat way to optimise this with the new iterator.